File: src/cli/engines/go/engine.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type Input struct {
	Action   string `json:"action"`
	Code     string `json:"code,omitempty"`
	Args     string `json:"args,omitempty"`
	Filename string `json:"filename,omitempty"`
}

type Output struct {
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
	ExitCode int    `json:"exit_code"`
	Result   string `json:"result"`
}

func writeOutput(o Output) {
	b, _ := json.Marshal(o)
	fmt.Println(string(b))
}

func execCode(code string) {
	tmpDir, err := os.MkdirTemp("", "klyron-go-*")
	if err != nil {
		writeOutput(Output{Stderr: err.Error(), ExitCode: 1})
		return
	}
	defer os.RemoveAll(tmpDir)

	srcPath := filepath.Join(tmpDir, "main.go")
	if err := os.WriteFile(srcPath, []byte(code), 0644); err != nil {
		writeOutput(Output{Stderr: err.Error(), ExitCode: 1})
		return
	}

	binPath := filepath.Join(tmpDir, "prog")
	cmd := exec.Command("go", "build", "-o", binPath, srcPath)
	if out, err := cmd.CombinedOutput(); err != nil {
		writeOutput(Output{Stderr: string(out), ExitCode: 1, Result: "Compilation failed"})
		return
	}

	cmd = exec.Command(binPath)
	out, err := cmd.CombinedOutput()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}
	}
	writeOutput(Output{Stdout: string(out), ExitCode: exitCode, Result: string(out)})
}

func evalExpr(expr string) {
	code := fmt.Sprintf(`package main
import "fmt"
func main() {
	fmt.Print(%s)
}`, expr)
	execCode(code)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		var input Input
		if err := json.Unmarshal([]byte(line), &input); err != nil {
			writeOutput(Output{Stderr: "Invalid JSON", ExitCode: 1})
			continue
		}

		switch input.Action {
		case "exec", "run":
			execCode(input.Code)
		case "eval":
			evalExpr(input.Code)
		case "ping", "":
			writeOutput(Output{Stdout: "pong", Result: "ok"})
		default:
			writeOutput(Output{Stderr: "Unknown action: " + input.Action, ExitCode: 1})
		}
	}
}

File: src/cli/engines/zig/engine.zig
const std = @import("std");
const json = std.json;
const fs = std.fs;
const process = std.process;

const Input = struct {
    action: []const u8,
    code: ?[]const u8 = null,
    args: ?[]const u8 = null,
    filename: ?[]const u8 = null,
};

const Output = struct {
    stdout: []const u8,
    stderr: []const u8,
    exit_code: i32,
    result: []const u8,
};

fn writeOutput(out: Output, writer: anytype) !void {
    const string = try json.stringifyAlloc(std.heap.page_allocator, out, .{});
    defer std.heap.page_allocator.free(string);
    try writer.print("{s}\n", .{string});
}

fn execCode(code: []const u8, writer: anytype) !void {
    const tmp_dir = "/tmp/klyron-zig-XXXXXX";
    var dir_buf: [64]u8 = undefined;
    @memcpy(&dir_buf, tmp_dir);
    const dir = try fs.cwd().makeOpenPath(std.os.mkdtemp(dir_buf[0..]), .{});
    defer fs.cwd().deleteTree(dir_buf[0..dir_buf.len]) catch {};

    const src_path = try dir.realpathAlloc(std.heap.page_allocator, "main.zig");
    defer std.heap.page_allocator.free(src_path);

    const bin_path = try dir.realpathAlloc(std.heap.page_allocator, "prog");
    defer std.heap.page_allocator.free(bin_path);

    try dir.writeFile("main.zig", code);

    const compile_result = try process.Child.run(.{
        .allocator = std.heap.page_allocator,
        .argv = &[_][]const u8{ "zig", "build-exe", src_path, "--name", "prog", "-O", "ReleaseFast" },
    });

    if (compile_result.term.Exited != 0) {
        try writeOutput(Output{
            .stdout = "",
            .stderr = compile_result.stderr,
            .exit_code = 1,
            .result = "Compilation failed",
        }, writer);
        return;
    }

    const run_result = try process.Child.run(.{
        .allocator = std.heap.page_allocator,
        .argv = &[_][]const u8{bin_path},
    });

    try writeOutput(Output{
        .stdout = run_result.stdout,
        .stderr = run_result.stderr,
        .exit_code = if (run_result.term.Exited) |code| @as(i32, @intCast(code)) else -1,
        .result = run_result.stdout,
    }, writer);
}

pub fn main() !void {
    const stdin = std.io.getStdIn().reader();
    const stdout = std.io.getStdOut().writer();
    var buf: [65536]u8 = undefined;

    while (try stdin.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        const trimmed = std.mem.trim(u8, line, " \n\r");
        if (trimmed.len == 0) continue;

        var parser = json.Parser.init(std.heap.page_allocator, false);
        defer parser.deinit();

        const input = parser.parse(trimmed) catch {
            try writeOutput(Output{ .stdout = "", .stderr = "Invalid JSON", .exit_code = 1, .result = "" }, stdout);
            continue;
        };

        const action = input.object.get("action") orelse continue;
        const action_str = action.string orelse "";

        if (std.mem.eql(u8, action_str, "exec") or std.mem.eql(u8, action_str, "run")) {
            const code = input.object.get("code") orelse continue;
            try execCode(code.string orelse "", stdout);
        } else if (std.mem.eql(u8, action_str, "ping") or action_str.len == 0) {
            try writeOutput(Output{ .stdout = "pong", .stderr = "", .exit_code = 0, .result = "ok" }, stdout);
        } else {
            try writeOutput(Output{ .stdout = "", .stderr = "Unknown action", .exit_code = 1, .result = "" }, stdout);
        }
    }
}
